crates/qasm2/src/lib.rs

```diff
@@ -0,0 +1,137 @@
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+#[macro_use]
+extern crate lazy_static;
+
+use pyo3::prelude::*;
+use pyo3::Python;
+
+mod bytecode;
+mod error;
+mod expr;
+mod lex;
+mod parse;
+
+/// Information about a custom instruction that Python space is able to construct to pass down to
+/// us.
+#[pyclass]
+#[derive(Clone)]
+pub struct CustomInstruction {
+    pub name: String,
+    pub num_params: usize,
+    pub num_qubits: usize,
+    pub builtin: bool,
+}
+
+#[pymethods]
+impl CustomInstruction {
+    #[new]
+    fn __new__(name: String, num_params: usize, num_qubits: usize, builtin: bool) -> Self {
+        Self {
+            name,
+            num_params,
+            num_qubits,
+            builtin,
+        }
+    }
+}
+
+/// Information about a custom classical function that should be defined in mathematical
+/// expressions.
+///
+/// The given `callable` must be a Python function that takes `num_params` floats, and returns a
+/// float.  The `name` is the identifier that refers to it in the OpenQASM 2 program.  This cannot
+/// clash with any defined gates.
+#[pyclass(text_signature = "(name, num_params, callable, /)")]
+#[derive(Clone)]
+pub struct CustomClassical {
+    pub name: String,
+    pub num_params: usize,
+    pub callable: PyObject,
+}
+
+#[pymethods]
+impl CustomClassical {
+    #[new]
+    fn __new__(name: String, num_params: usize, callable: PyObject) -> Self {
+        Self {
+            name,
+            num_params,
+            callable,
+        }
+    }
+}
+
+/// Create a bytecode iterable from a string containing an OpenQASM 2 program.  The iterable will
+/// lex and parse the source lazily; evaluating OpenQASM 2 statements as required, without loading
+/// the entire token and parse tree into memory at once.
+#[pyfunction]
+fn bytecode_from_string(
+    _py: Python<'_>,
+    string: String,
+    include_path: Vec<std::ffi::OsString>,
+    custom_instructions: Vec<CustomInstruction>,
+    custom_classical: Vec<CustomClassical>,
+    strict: bool,
+) -> PyResult<bytecode::BytecodeIterator> {
+    bytecode::BytecodeIterator::new(
+        lex::TokenStream::from_string(string, strict),
+        include_path.iter().map(|x| x.into()).collect(),
+        &custom_instructions,
+        &custom_classical,
+        strict,
+    )
+}
+
+/// Create a bytecode iterable from a path to a file containing an OpenQASM 2 program.  The
+/// iterable will lex and parse the source lazily; evaluating OpenQASM 2 statements as required,
+/// without loading the entire token and parse tree into memory at once.
+#[pyfunction]
+fn bytecode_from_file(
+    _py: Python<'_>,
```

Comment:
Ditto that it should be fine to leave this off

Reply:
Done in 203c7bb9e.